File: Daily/28/02/1092.py
class Solution:
    def shortestCommonSupersequence(self, str1: str, str2: str) -> str:
         m = len(str1)
         n = len(str2)
         
         dp = [[0] * (n + 1) for _ in range(m + 1)]
         
         for i in range( 1, m + 1):
             for j in range( 1, n + 1):
                 if str1[i - 1] == str2[j - 1]:
                     dp[i][j] = dp[i -1][j -1] + 1
                 else:
                     dp[i][j] = max(dp[i - 1][j],dp[i][j - 1])
                     
         i = m 
         j = n
         result = ""
         
         while i > 0 or j > 0:
             if i > 0 and j > 0 and str1[i -1] == str2[j -1]:
                 result = str1[i - 1] + result
                 i -= 1
                 j -= 1
             elif i > 0 and (j == 0 or dp[i -1][j] > dp[i][j -1]):
                 result = str1[i - 1] + result
                 i -= 1
             else:
                 result = str2[j -1 ] + result
                 j -= 1
         return result 

File: Daily/28/1092.java
class Solution {
    public String shortestCommonSupersequence(String str1, String str2) {
        int m = str1.length();
        int n = str2.length();

        // 1. Find LCS using dynamic programming
        int[][] dp = new int[m + 1][n + 1];

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (str1.charAt(i - 1) == str2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }

        // 2. Construct the SCS
        int i = m;
        int j = n;
        StringBuilder result = new StringBuilder();

        while (i > 0 || j > 0) {
            if (i > 0 && j > 0 && str1.charAt(i - 1) == str2.charAt(j - 1)) {
                result.insert(0, str1.charAt(i - 1));
                i--;
                j--;
            } else if (i > 0 && (j == 0 || dp[i - 1][j] >= dp[i][j - 1])) {
                result.insert(0, str1.charAt(i - 1));
                i--;
            } else {
                result.insert(0, str2.charAt(j - 1));
                j--;
            }
        }

        return result.toString();
    }
}